java/code/src/com/suse/cloud/CloudPaygManager.java

```diff
@@ -155,226 +133,77 @@ public boolean checkRefreshCache(boolean force) {
-     */
-    public void setCloudProvider(CloudProvider providerIn) {
-        cloudProvider = providerIn;
-        cacheTime = Instant.now();
-    }
-
-    /**
-     * Only for testing - overwrite the detected isPaygInstance
-     * @param isPaygInstanceIn is payg instance
-     */
-    public void setPaygInstance(boolean isPaygInstanceIn) {
-        isPaygInstance = isPaygInstanceIn;
-        cacheTime = Instant.now();
-    }
-
-    /**
-     * Only for testing - overwrite the detected hasSCCCredentials
-     * @param hasSCCCredentialsIn has scc credentials
-     */
-    public void setHasSCCCredentials(boolean hasSCCCredentialsIn) {
-        hasSCCCredentials = hasSCCCredentialsIn;
-        cacheTime = Instant.now();
-    }
-
-    /**
-     * Only for testing - overwrite the detected isCompliant
-     * @param isCompliantIn is system compliant
-     */
-    public void setCompliant(boolean isCompliantIn) {
-        isCompliant = isCompliantIn;
-        cacheTime = Instant.now();
-    }
-
-    private CloudProvider detectCloudProvider() {
-        if (isFileExecutable("/usr/bin/ec2metadata")) {
-            return CloudProvider.AWS;
-        }
-        else if (isFileExecutable("/usr/bin/azuremetadata")) {
-            return CloudProvider.AZURE;
-        }
-        else if (isFileExecutable("/usr/bin/gcemetadata")) {
-            return CloudProvider.GCE;
-        }
-        return CloudProvider.None;
-    }
-
-    private boolean detectPaygInstance() {
-        if (!isFileExecutable("/usr/bin/instance-flavor-check")) {
-            return false;
-        }
-
-        try {
-            return "PAYG".equals(getInstanceType());
-        }
-        catch (ExecutionException ex) {
-            LOG.error("Unable to identify the instance type. Fallback to BYOS.", ex);
-            return false;
-        }
-    }
-
     private boolean detectHasSCCCredentials() {
         return CredentialsFactory.listSCCCredentials().stream()
                 .anyMatch(c -> mgr.isSCCCredentials(c));
     }
 
     private boolean detectIsCompliant() {
-        // do not use isPaygInstance() to prevent a loop
-        if (BooleanUtils.isFalse(isPaygInstance)) {
+        // Parse the compliance info from the external json file, if available
+        complainceInfo = getInstanceComplianceInfo();
+
+        // If it's not payg, it's always compliant
+        if (!complainceInfo.isPaygInstance()) {
             return true;
         }
 
+        // If the external compliance tool report it's not ok, it cannot be compliant
+        if (!complainceInfo.isCompliant()) {
+            LOG.error("The instance has been reported as non-compliant.");
+            if (complainceInfo.isAnyPackageModified()) {
+                LOG.error("Some packages have been modified and cannot be trusted.");
+            }
+            if (complainceInfo.isBillingServiceRunning()) {
+                LOG.error("The CSP Billing Adapter service is not active.");
+            }
```

Comment:
The new BillingAdapterStatus should be used here as well.

Reply:
Done and also fixed the check (it should have been negated).